20240927-multiple-object-deletion.md

```diff
@@ -0,0 +1,163 @@
+    RequestHeader header = 15;
+
+    bytes bucket = 1;
+    bool bypass_governance_retention = 2;
+    bool quiet = 3;
+    repeated DeleteObjectsRequestItem items = 4;
+}
+
+message DeleteObjectsRequestItem {
+    bytes encrypted_object_key = 1;
+    bytes object_version = 2;
+}
+
+message DeleteObjectsResponse {
+	repeated DeleteObjectsResponseDeletedItem deleted = 1;
+	repeated DeleteObjectsResponseErrorItem errors = 2;
+}
+
+message DeleteObjectsResponseDeletedItem {
+    bytes encrypted_object_key = 1;
+    bytes deleted_object_version = 2;
+    bytes delete_marker_version = 3;
+}
+
+message DeleteObjectsResponseErrorItem {
+    int32 code = 1;
+    bytes encrypted_object_key = 2;
+    bytes object_version = 3;
+}
+```
+
+## Disclaimers
+
+### Anti-goals
+
+### Alternatives considered
+
+An alternative to the design described in this document is to implement a separate rate limit specifically for bulk object deletion operations. However, bulk deletion operations could still fail partway through due to hitting the bulk deletion-specific rate limit, even though other operations would remain unaffected. In addition, unnecessary bandwidth consumption would remain an issue, as each object deletion would still require a separate request.
+
+### Open questions
+
+## Reminders
+
+### Security / Privacy
+
+The same security and privacy considerations that apply to the existing single-object deletion implementation must apply to the multiple object deletion implementation. We must ensure that the client is unable to delete any objects for which deletion is not permitted by its API key. Additionally, Object Lock restrictions must be obeyed.
+
+### Observability
+
+We plan to use the same observability practices that we currently use for existing metainfo RPC service endpoints. Each request triggers an Eventkit event containing request metadata, and Monkit instrumentation is added to each non-trivial method involved in the processing of the request.
+
+Customers may enable [bucket logging](https://storj.dev/dcs/buckets/bucket-logging) in order to view access logs for S3 requests - including those for `DeleteObjects` operations - that are sent to our S3-compatible gateway.
+
+### Test plan
+
+Tests for this design should confirm all of the following statements:
+
+- If the bucket is versioned and the version ID of an object isn't specified, a delete marker will be placed at the object's location, and no deletions will occur. This will be the case even if there isn't an object version at the location.
+- If the bucket has versioning suspended and the version ID of an object isn't specified, a delete marker will be placed at the object's location. If there's an object version at the location, it will be deleted.
+- If the bucket is unversioned and the version ID of an object isn't specified, the object will be deleted without the insertion of a delete marker at its location. If no object exists at the location, an error will be returned.
```

Comment:
doesn't S3 just say it was deleted if the object doesn't exist?

> Amazon S3 performs a delete operation and returns the result of that delete, success or failure, in the response. Note that if the object specified in the request is not found, Amazon S3 returns the result as deleted.

https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#deleteObject-property

Reply:
Thank you for the correction; amended.